Mempool package tracking

sdaftuar opened the pull request:
This is a preparatory pull to try to make reviewing #6557 easier.

In #6557, I added tracking packages of transactions to the mempool (for each tx, we track all the "descendant" transactions that depend on that tx), in order to make the mempool limiting code more effective.  This PR is a standalone implementation of mempool descendant tracking, including the policy limits on transaction chains (limiting ancestors and descendants) proposed in #6557.

I'll rebase that pull off these commits assuming we can agree on this approach.

ashleyholman committed:
TxMemPool: Change mapTx to a boost::multi_index_container

Indexes on:
- Tx Hash
- Fee Rate (fee-per-kb)

sdaftuar committed:
Track transaction packages in CTxMemPoolEntry

Associate with each CTxMemPoolEntry all the size/fees of descendant
mempool transactions.  Sort mempool by max(feerate of entry, feerate
of descendants).  Update statistics on-the-fly as transactions enter
or leave the mempool.

Also add ancestor and descendant limiting, so that transactions can
be rejected if the number or size of unconfirmed ancestors exceeds
a target, or if adding a transaction would cause some other mempool
entry to have too many (or too large) a set of unconfirmed in-
mempool descendants.

sdaftuar commented:
@sipa  I did end up reworking the tracking of in-mempool parents/children to use iterators rather than hashes as you had suggested (which I never pushed up to #6557).

btcdrak commented:
@laanwj looks like travis crapped out on one of the jobs and needs restarting.

sdaftuar commented:
@btcdrak Actually I think this is some kind of problem with the unit test code -- not sure why it fails to compile only in that one job but I was just able to reproduce locally.

sdaftuar commented:
@btcdrak Fixed and pushed back up, travis is happy now...

sdaftuar commented on src/main.h:
```diff
@@ -466,5 +474,7 @@ static const unsigned int REJECT_HIGHFEE = 0x100;
 /** Transaction conflicts with a transaction already known */
 static const unsigned int REJECT_CONFLICT = 0x102;
+/** Transaction would result in too long in-mempool chain */
+static const unsigned int REJECT_LONGCHAIN = 0x103;
```
Will move this to not be an internal code (so we send a reject message back).

petertodd commented on src/main.cpp:
```diff
@@ -921,6 +921,17 @@ bool AcceptToMemoryPool(CTxMemPool& pool, CValidationState &state, const CTransa
+        // Calculate in-mempool ancestors, up to a limit.
+        CTxMemPool::setEntries setAncestors;
+        size_t nLimitAncestors = GetArg("-limitancestorcount", DEFAULT_ANCESTOR_LIMIT);
+        size_t nLimitAncestorSize = GetArg("-limitancestorsize", DEFAULT_ANCESTOR_SIZE_LIMIT)*1000;
```
As the idea is the ancestor size is set such that the whole string of transactions could be included in a single block, we really want this to be a less than the max block size, as there's always overhead to consider. (e.g. the coinbase, soft-limit, etc)

I'd suggest we set this to 900KB for a 100KB buffer - plenty even in the case of p2pool's large max 50KB coinbase. Another option might be to set it based on the soft-limit - default 750KB - with another 100KB of overhad buffer.

petertodd commented:
In general, it'd be good to think about separating the reorg case entirely from the main mempool codebase. For instance, keep a simple linear list of reorged transactions, and after a reorg attempt to add them back the mempool one-by-one. This separate code could also handle cases where we might want to remine transactions that we otherwise wouldn't, as a "goodwill" gesture to reduce the impact of large reorgs.

petertodd commented on src/txmempool.cpp:
```diff
@@ -48,6 +53,246 @@ CTxMemPoolEntry::GetPriority(unsigned int currentHeight) const
+{
+    // Track the number of entries (outside setExclude) that we'd need to visit
+    // (will bail out if it exceeds maxDescendantsToVisit)
+    int nChildrenToVisit = 0; 
```
trailing whitespace

pstratem commented:
Needs a comment defining ancestor/descendant.

petertodd commented on src/txmempool.cpp:
```diff
@@ -48,6 +53,246 @@ CTxMemPoolEntry::GetPriority(unsigned int currentHeight) const
+
+    while (!parentHashes.empty()) {
+        setAncestors.insert(parentHashes.begin(), parentHashes.end());
+        setEntries stageParentSet; 
```
trailing whitespace

petertodd commented on src/txmempool.cpp:
```diff
@@ -48,6 +53,246 @@ CTxMemPoolEntry::GetPriority(unsigned int currentHeight) const
+                    errString = strprintf("too many unconfirmed ancestors [limit: %u]", limitAncestorCount);
+                    return false;
+                }
+            }    
```
trailing whitespace

petertodd commented on src/txmempool.cpp:
```diff
@@ -89,34 +334,106 @@ void CTxMemPool::AddTransactionsUpdated(unsigned int n)
+    minerPolicyEstimator->removeTx(hash);
+}
+
+// Calculates descendants of hash that are not already in setDescendants, and adds to 
```
trailing whitespace

petertodd commented on src/txmempool.cpp:
```diff
@@ -247,19 +554,23 @@ void CTxMemPool::check(const CCoinsViewCache *pcoins) const
-        innerUsage += it->second.DynamicMemoryUsage();
-        const CTransaction& tx = it->second.GetTx();
+        checkTotal += it->GetTxSize();
+        innerUsage += it->DynamicMemoryUsage(); 
```
trailing whitespace

petertodd commented on src/txmempool.h:
```diff
@@ -138,6 +363,29 @@ class CTxMemPool
     void ClearPrioritisation(const uint256 hash);
 
+public:
+    void RemoveStaged(std::set<uint256>& stage);
```
Need to document what RemoveStaged() is.

petertodd commented on src/txmempool.h:
```diff
@@ -138,6 +363,29 @@ class CTxMemPool
+    /** When adding transactions from a disconnected block back to the mempool,
+     *  new mempool entries may have children in the mempool (which is generally
+     *  not the case when otherwise adding transactions).
+     *  UpdateTransactionsFromBlock will find child transactions and update the
```
Would prefer if comments like this are written as "Name()" rather than just "Name" to make it clear what's a variable and what's a function.

petertodd commented on src/txmempool.h:
```diff
@@ -169,6 +417,48 @@ class CTxMemPool
+    /** UpdateForDescendants is used by UpdateTransactionsFromBlock to update
+     *  the descendants for a single transaction that has been added to the
+     *  mempool but may have child transactions in the mempool, eg during a
+     *  chain reorg.  setExclude is the set of descendant transactions in the
```
Double-space after periods? Obvs you're actually Satoshi.

petertodd commented on src/txmempool.h:
```diff
@@ -138,6 +363,29 @@ class CTxMemPool
+     *  descendant state for each transaction in hashesToUpdate (excluding any
+     *  child transactions present in hashesToUpdate, which are already accounted
+     *  for).
+     */
```
This comment doesn't make it clear whether or not transactions in hashesToUpdate are or are not already in the mempool. :)

sipa commented:
Concept ACK.

petertodd commented:
utACK, modulo nits.

jonasschnelli commented:
Concept ACK / code review ACK.
Passes gitian/osx/debian build/unit-tests/rpc-tests.

petertodd commented:
Tested with my FSS and Full RBF stress tests and -checkmempool, no unexpected failures. (this pull-req of course doesn't enable any RBF behavior, so replacements failed!)

dcousens commented:
concept ACK

dcousens commented on src/main.cpp:
```diff
@@ -921,6 +921,17 @@ bool AcceptToMemoryPool(CTxMemPool& pool, CValidationState &state, const CTransa
 
+        // Calculate in-mempool ancestors, up to a limit.
+        CTxMemPool::setEntries setAncestors;
+        size_t nLimitAncestors = GetArg("-limitancestorcount", DEFAULT_ANCESTOR_LIMIT);
```
Mixing `GetArg` in mempool code feels a bit odd.

sdaftuar replied:
I agree but is there a better way to do this?  I think I'm following the common practice in the code base, though I look forward to a future refactoring when we have better encapsulation.

dcousens commented on src/txmempool.cpp:
```diff
@@ -48,6 +53,246 @@ CTxMemPoolEntry::GetPriority(unsigned int currentHeight) const
+
+        setEntries entriesToAdd;
+        BOOST_FOREACH(const txiter cit, stageEntries) {
+            if (cit->IsDirty()) {
```
Dirty implies something changed typically,  perhaps instead change this to `ShouldSkip`?

dcousens commented on src/txmempool.cpp:
```diff
@@ -48,6 +53,246 @@ CTxMemPoolEntry::GetPriority(unsigned int currentHeight) const
+        // include them, and update their setMemPoolParents to include this tx.
+        for (; iter != mapNextTx.end() && iter->first.hash == hash; ++iter) {
+            const uint256 &childHash = iter->second.ptx->GetHash();
+            txiter childIter = mapTx.find(childHash);
```
Do we need to check `childIter == mapTx.end()`?

sdaftuar replied:
I'll add an assert to make clear that is not possible.

dcousens commented on src/txmempool.cpp:
```diff
@@ -48,6 +53,246 @@ CTxMemPoolEntry::GetPriority(unsigned int currentHeight) const
+{
+    // For each entry, walk back all ancestors and decrement size associated with this
+    // transaction
+    uint64_t nNoLimit = std::numeric_limits<uint64_t>::max();
```
const?

dcousens commented on src/txmempool.cpp:
```diff
@@ -48,6 +53,246 @@ CTxMemPoolEntry::GetPriority(unsigned int currentHeight) const
+{
+    nCountWithDescendants=0;
+    nSizeWithDescendants=nTxSize;
+    nFeesWithDescendants=nFee;
```
Spacing between operators

dcousens commented on src/txmempool.cpp:
```diff
@@ -48,6 +53,246 @@ CTxMemPoolEntry::GetPriority(unsigned int currentHeight) const
+    }
+    int64_t updateCount = (add ? 1 : -1);
+    int64_t updateSize = updateCount * it->GetTxSize();
+    CAmount updateFee = updateCount * it->GetFee();
```
nit: `const` on these?

dcousens commented on src/txmempool.cpp:
```diff
@@ -48,6 +53,246 @@ CTxMemPoolEntry::GetPriority(unsigned int currentHeight) const
+// TODO: pass a txiter instead?
+void CTxMemPool::UpdateChildrenForRemoval(const uint256 &hash)
+{
+    txiter it = mapTx.find(hash);
```
Check/assert `it == mapTx.end()`?

dcousens commented on src/txmempool.cpp:
```diff
@@ -48,6 +53,246 @@ CTxMemPoolEntry::GetPriority(unsigned int currentHeight) const
+    uint64_t nNoLimit = std::numeric_limits<uint64_t>::max();
+    BOOST_FOREACH(const uint256& removeHash, hashesToRemove) {
+        setEntries setAncestors;
+        const CTxMemPoolEntry &entry = *mapTx.find(removeHash);
```
Check/assert `it == mapTx.end()`, especially before de-reference

dcousens commented on src/txmempool.cpp:
```diff
@@ -89,34 +334,106 @@ void CTxMemPool::AddTransactionsUpdated(unsigned int n)
+        setDescendants.insert(stage.begin(), stage.end());
+        std::set<uint256> setNext;
+        BOOST_FOREACH(const uint256 &stagehash, stage) {
+            indexed_transaction_set::iterator it = mapTx.find(stagehash);
```
Check/assert `it == mapTx.end()`?
Would only be possible on the first `stagehash` AFAIK

dcousens commented on src/txmempool.cpp:
```diff
@@ -125,34 +442,21 @@ void CTxMemPool::remove(const CTransaction &origTx, std::list<CTransaction>& rem
+            setAllRemoves = txToRemove;
+        }
+        BOOST_FOREACH(const uint256& hash, setAllRemoves) {
+            removed.push_back(mapTx.find(hash)->GetTx());
```
Again,  assumption of `hash` in `mapTx`,  worth asserting?

dcousens commented on src/txmempool.cpp:
```diff
@@ -271,8 +582,34 @@ void CTxMemPool::check(const CCoinsViewCache *pcoins) const
+        int64_t childSizes = 0;
+        CAmount childFees = 0;
+        for (; iter != mapNextTx.end() && iter->first.hash == it->GetTx().GetHash(); ++iter) {
+            txiter childit = mapTx.find(iter->second.ptx->GetHash());
```
Again, `find` not checked,  if this assumption can hold true,  can it be documented?

sdaftuar replied:
mapNextTx contains exactly the in-mempool children of a given transaction, so this code should be safe.  I'll add an assert to make the assumption clearer (the semantics of mapNextTx weren't changed by this pull).

dcousens commented on src/txmempool.cpp:
```diff
@@ -247,19 +554,23 @@ void CTxMemPool::check(const CCoinsViewCache *pcoins) const
+        checkTotal += it->GetTxSize();
+        innerUsage += it->DynamicMemoryUsage(); 
+        const CTransaction& tx = it->GetTx();
+        const TxLinks &links = mapLinks.find(it)->second;
```
Is `mapLinks` in sync with `mapTx`?

sdaftuar replied:
Yes -- I'll be adding some additional asserts to make this more clear.

sdaftuar commented:
Thanks everyone for reviewing!  I've pushed up a series of commits to address everyone's feedback; I believe all comments should have been addressed.

These cleanups need to be squashed, and this pull now needs to be rebased (a merge conflict crept in now that I'm outputting additional information in getrawmempool).  Reviewers -- please let me know if you prefer I leave these commits unsquashed/unrebased so you can review the changes.  (In the absence of any expressed preferences, I'll plan to squash/rebase in a day or two so that this can become mergeable.)

@petertodd I added the extra information to the getrawmempool RPC call (good idea!).  Now that the RPC call is there, I realized that adding an RPC test that exercises the new limits is a good idea, so I've started work on that, but it's not yet complete.

dcousens commented:
utACK,  thanks @sdaftuar !

morcos commented on src/txmempool.h:
```diff
@@ -99,8 +285,43 @@ class CTxMemPool
-    std::map<uint256, CTxMemPoolEntry> mapTx;
+    indexed_transaction_set mapTx;
+    typedef indexed_transaction_set::iterator txiter;
+    struct CompareIteratorByHash {
```
Can all these new members from here through to `UpdateChild` be made private?

morcos commented on src/txmempool.cpp:
```diff
@@ -429,5 +758,58 @@ bool CCoinsViewMemPool::HaveCoins(const uint256 &txid) const {
+}
+
+bool CTxMemPool::addUnchecked(const uint256&hash, const CTxMemPoolEntry &entry, bool fCurrentEstimate)
+{
```
needs LOCK(cs)?

morcos commented on src/txmempool.cpp:
```diff
@@ -429,5 +758,58 @@ bool CCoinsViewMemPool::HaveCoins(const uint256 &txid) const {
+    return memusage::MallocUsage(sizeof(CTxMemPoolEntry) + 9 * sizeof(void*)) * mapTx.size() + memusage::DynamicUsage(mapNextTx) + memusage::DynamicUsage(mapDeltas) + memusage::DynamicUsage(mapLinks) + cachedInnerUsage;
+}
+
+void CTxMemPool::RemoveStaged(setEntries &stage) {
```
Needs LOCK(cs) if ever called from somewhere else.

TheBlueMatt replied:
You can also AssertLockHeld(cs), otherwise, to make things more readable.

On September 16, 2015 11:05:34 AM CDT, Alex Morcos <notifications@github.com> wrote:
>> @@ -429,5 +758,58 @@ bool CCoinsViewMemPool::HaveCoins(const uint256
>&txid) const {
>>  
>>  size_t CTxMemPool::DynamicMemoryUsage() const {
>>      LOCK(cs);
>> -    return memusage::DynamicUsage(mapTx) +
>memusage::DynamicUsage(mapNextTx) + memusage::DynamicUsage(mapDeltas) +
>cachedInnerUsage;
>> +    // Estimate the overhead of mapTx to be 9 pointers + an
>allocation, as no exact formula for boost::multi_index_contained is
>implemented.
>> +    return memusage::MallocUsage(sizeof(CTxMemPoolEntry) + 9 *
>sizeof(void*)) * mapTx.size() + memusage::DynamicUsage(mapNextTx) +
>memusage::DynamicUsage(mapDeltas) + memusage::DynamicUsage(mapLinks) +
>cachedInnerUsage;
>> +}
>> +
>> +void CTxMemPool::RemoveStaged(setEntries &stage) {
>
>Needs LOCK(cs) if ever called from somewhere else.
>
>---
>Reply to this email directly or view it on GitHub:
>https://github.com/bitcoin/bitcoin/pull/6654/files#r39649424

morcos commented on src/txmempool.cpp:
```diff
@@ -48,6 +46,243 @@ CTxMemPoolEntry::GetPriority(unsigned int currentHeight) const
+// for each entry, look for descendants that are outside hashesToUpdate, and
+// add fee/size information for such descendants to the parent.
+void CTxMemPool::UpdateTransactionsFromBlock(const std::vector<uint256> &vHashesToUpdate)
+{
```
Needs LOCK(cs)

morcos commented on src/txmempool.h:
```diff
@@ -99,8 +285,43 @@ class CTxMemPool
+            return a->GetTx().GetHash() < b->GetTx().GetHash();
+        }
+    };
+    typedef std::set<txiter, CompareIteratorByHash> setEntries;
```
oops, `setEntries` needs to be public, the rest can be private though.

TheBlueMatt commented:
My computer's battery died, but in txmempool.h, the definition of txiter as "boost::multi_index_container::iterator" is "implementation defined" according to boost's docs. In the latest versions it is defined to an equivalent of "boost::multi_index_container::nth_index<0>::type::iterator", so you should use that.

On September 16, 2015 8:44:26 AM CDT, Suhas Daftuar <notifications@github.com> wrote:
>Thanks everyone for reviewing!  I've pushed up a series of commits to
>address everyone's feedback; I believe all comments should have been
>addressed.
>
>These cleanups need to be squashed, and this pull now needs to be
>rebased (a merge conflict crept in now that I'm outputting additional
>information in getrawmempool).  Reviewers -- please let me know if you
>prefer I leave these commits unsquashed/unrebased so you can review the
>changes.  (In the absence of any expressed preferences, I'll plan to
>squash/rebase in a day or two so that this can become mergeable.)
>
>@petertodd I added the extra information to the getrawmempool RPC call
>(good idea!).  Now that the RPC call is there, I realized that adding
>an RPC test that exercises the new limits is a good idea, so I've
>started work on that, but it's not yet complete.
>
>---
>Reply to this email directly or view it on GitHub:
>https://github.com/bitcoin/bitcoin/pull/6654#issuecomment-140746438

dcousens commented:
@sdaftuar needs rebase.

sdaftuar commented:
I addressed the latest comments from @morcos and @TheBlueMatt, added an rpc test (it only tests the ancestor/descendant length limits, not the size limits, nor the handling of reorgs -- so that still can be improved), and then squashed everything down and rebased on master to get rid of the merge conflict in ```rpcblockchain.cpp```.

Mirobit commented:
Does anyone else thing the ancestor limit is way to generous? In what other case than spamming do txs have 100 ancestors in mempool? Shouldn't 5 be enough?

petertodd commented:
@Mirobit The limits are there to limit computational costs in determining things like dependent fees/size etc. There's no need to set them low unless the algorithms take a long time to compute those sums; the data structures in the mempool are mainly pointer following so it's fairly fast.

laanwj commented:
utACK

sipa commented:
My node running 5add7a7 with -checkmempool crashed. Last debug.log message was:
```
2015-09-22 13:06:19 - Disconnect block: 90.04ms
```

morcos commented:
@sipa can you give us some more details?  were you running will all -debug options?  for instance I assume you were not running with estimatefee debugging?   it helps to narrow down where it crashed based on what didn't get printed to the debug log.
